docs/RationaleRTL-SV.md

```diff
@@ -105,6 +105,114 @@ declarations in limited ways:
  - Interface signals are allowed to be zero bits wide.  They are dropped from
    Verilog emission.
 
+### Endianness: operand ordering and internal representation
+
+Certain operations require ordering to be defined (i.e. `rtl.concat`,
+`rtl.array_concat`, and `rtl.array_create`). There are two places where this
+is relevant: in the MLIR assembly and in the MLIR C++ model.
+
+In MLIR assembly, operands are always listed MSB to LSB (big endian style):
+
+```mlir
+%msb = rtl.constant (0xEF : i8) : i8
+%mid = rtl.constant (0x7 : i4) : i4
+%lsb = rtl.constant (0xA018 : i16) : i16
+%result = rtl.concat %msb, %mid, %lsb : (i8, i4, i16) -> i28
+// %result is 0xEF7A018
+```
+
+**Note**: Integers are always big endian in both bit and byte ordering.
+Operand ordering for `rtl.concat` was chosen to be consistent with simply
+abutting them in lexical order.
+
+```mlir
+%1 = rtl.constant (0x1 : i4) : i4
+%2 = rtl.constant (0x2 : i4) : i4
+%3 = rtl.constant (0x3 : i4) : i4
+%arr123 = rtl.array_create %1, %2, %3 : i4
+// %arr123[0] = 0x3
+// %arr123[1] = 0x2
+// %arr123[2] = 0x1
+
+%arr456 = ... // {0x4, 0x5, 0x6}
+%arr78  = ... // {0x7, 0x8}
+%arr = rtl.array_concat %arr123, %arr456, %arr78 : !rtl.array<3 x i4>, !rtl.array<3 x i4>, !rtl.array<2 x i4>
+// %arr[0] = 0x6
+// %arr[1] = 0x5
+// %arr[2] = 0x4
+// %arr[3] = 0x3
+// %arr[4] = 0x2
+// %arr[5] = 0x1
+```
+
+**Note**: This ordering scheme is unintuitive for anyone expecting C
+array-like ordering. In a C, arrays are laid out with index 0 as the least
+significant value. In the CIRCT _model_ (assembly and C++), it is the
+opposite -- the most significant value is located at the zero index. The
+produced hardware, however, _is_ in C-style order as operands are printed (in
+SystemVerilog) in the same lexical order as the assembly. Since SystemVerilog
+array concatenation and array creation are listed in MSB to LSB
+left-to-right, the ordering is reversed again.
+
+In the CIRCT C++ model, lists of values are in lexical order. That is, index
+zero of a list is the leftmost operand in assembly, which is the most
+significant value.
+
+```cpp
+ConcatOp result = builder.create<ConcatOp>(..., {msb, lsb});
+// Is equivalent to the above integer concatenation example.
+ArrayConcatOp arr = builder.create<ArrayConcatOp>(..., {arr123, arr456});
+// Is equivalent to the above array example.
+```
+
+### Array slicing and indexing
+
+Should `array_get %arr123[%1]` (wherein `%1` is dynamically zero) result in
```

Comment:
Indexing and slicing also needs to be defined crisply.

Reply:
I should also note that I prefer the way it is today since having the index semantics match the produced hardware will produce what people think it will. IMO we should just note that `get` is not an index into the operand list, it is an index into the array in LSB = 0 index order.

Reply:
I suspect some of the confusion and objections in today's meeting were from conflating the operand list in the operation creating the list with the runtime indexing. (I certainly didn't fully grok the difference until after the meeting.) I would hazard a guess that few of us are used to reasoning about the compiler rather than the code it produces.